src/client/utils/hooks/useSelectedInsuranceTypes.ts

```diff
@@ -36,34 +44,57 @@ const validateInsuranceTypes = (
       ALL_CONTRACT_TYPES.includes((type as unknown) as TypeOfContract),
   ) as Array<InsuranceType>
 
-export const getSelectedInsuranceTypes = () =>
-  deserializeSearchParams(new URLSearchParams(window.location.search))
+const SESSION_STORAGE_PREFIX = 'HEDVIG_'
+const getStorageKey = (quoteCartId: string) =>
+  `${SESSION_STORAGE_PREFIX}${quoteCartId}`
 
 export const useSelectedInsuranceTypes = () => {
-  const location = useLocation()
-  const history = useHistory()
+  const { id: quoteCartId } = useParams()
+  const storageKey = getStorageKey(quoteCartId)
 
-  const searchParams = useMemo(() => new URLSearchParams(location.search), [
-    location.search,
-  ])
-  const insuranceTypes = useMemo(() => deserializeSearchParams(searchParams), [
-    searchParams,
-  ])
+  const [insuranceTypes, setInsuranceTypes] = useState(() => {
+    const raw = window.sessionStorage.getItem(storageKey)
+    if (raw) {
+      const storedInsuranceTypes = deserializeInsuranceTypes(raw)
+      if (storedInsuranceTypes.length > 0) return storedInsuranceTypes
+    }
 
-  const changeSelectedInsuranceTypes = useCallback(
-    (newTypes: Array<InsuranceType | TypeOfContract>) => {
-      searchParams.delete(SEARCH_PARAM_NAME)
+    const searchParams = new URLSearchParams(location.search)
+    const initialInsuranceTypes = deserializeSearchParams(searchParams)
+    return initialInsuranceTypes
+  })
 
-      for (const type of newTypes) {
-        searchParams.append(SEARCH_PARAM_NAME, type)
-      }
+  const changeSelectedInsuranceTypes = useCallback(
+    (newTypes: InsuranceType[] | TypeOfContract[]) => {
+      window.sessionStorage.setItem(
+        storageKey,
+        serializeInsuranceTypes(newTypes),
+      )
 
-      history.replace({
-        search: searchParams.toString(),
-      })
+      const event = new CustomEvent<InsuranceTypesEventData>(
+        INSURANCE_TYPES_EVENT,
+        {
+          detail: newTypes,
+          bubbles: true,
+        },
+      )
+      document.dispatchEvent(event)
     },
-    [searchParams, history],
+    [storageKey],
   )
 
+  useEffect(() => {
+    const handler = (event: CustomEvent<InsuranceTypesEventData>) => {
+      setInsuranceTypes(event.detail)
+    }
+
+    window.addEventListener(INSURANCE_TYPES_EVENT, handler as EventListener)
```

Comment:
Good point! We wanted to use the built-in "storage" event but it is only sent to iframes in the same page. Not to subscribers in the same context for some odd reason...

https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event

A better solution would be to wrap components in a context. However, that would require significant changes.

Perhaps we should look into something like Jotai to more easily handle these things!

Reply:
I still don't understand this...
Why not call `setInsuranceTypes(event.detail)` instead of dispatching an event which we listen to and execute `setInsuranceTypes(event.detail)`?